patterns.md

```diff
@@ -0,0 +1,66 @@
+# Constants in patterns
+
+Constants that pass the [structural equality check][struct-eq] can be used in patterns:
+```rust
+const FOO: i32 = 13;
+
+fn is_foo(x: i32) -> bool {
+  match x {
+    FOO => true,
+    _ => false,
+  }
+}
+```
+However, that check has some loopholes, so e.g. `&T` can be used in a pattern no matter the `T`.
+
+Any reference type const-pattern is compiled by [calling `PartialEq::eq`][compile-partial-eq] to compare the subject of the `match` with the constant. Const-patterns with other types (enum, struct, tuple, array) are treated as if the constant was inlined as a pattern (and the usual `match` tree is constructed).
+Note that this is in contradiction with what has been RFC'd, which specifies that we should always call `PartialEq::eq`.
```

Comment:
Would be good to link to the RFC here. I assume that's https://github.com/rust-lang/rfcs/blob/master/text/1445-restrict-constants-in-patterns.md but so far I have not found where it says that we should use `PartialEq`.

Reply:
The RFC calls this "semantic equality" and makes some preparation for maybe switching to it in the future, but it dosn't actually say that we want to adopt semantic equality. That's my reading, anyway.

Reply:
It's a bit spread around, but https://github.com/rust-lang/rfcs/blob/master/text/1445-restrict-constants-in-patterns.md#disadvantage-weakened-abstraction-bounary is probably the best entry point

Reply:
But later the RFC explicitly says

> The goal of this RFC is not to decide between semantic and structural equality. Rather, the goal is to restrict pattern matching to that subset of types where the two variants behave roughly the same.

Reply:
It's very explicit in the [detailed design section](https://github.com/rust-lang/rfcs/blob/master/text/1445-restrict-constants-in-patterns.md#the-structural-match-attribute)

>  When converting a constant value into a pattern, if the constant is of struct or enum type, we will check whether this attribute is present on the struct -- if so, we will convert the value as we do today. If not, we will report an error that the struct/enum value cannot be used in a pattern.

where "as we do today" is "convert to an equivalent pattern"

Reply:
I don't think that is more explicit that "The goal of this RFC is not to decide between semantic and structural equality".

That text you quoted does not say anywhere that we should use semantic equality for anything.

Reply:
I adjusted the text to reflect what I think the RFC says.